lib/render/FlowRenderer.cpp

```diff
@@ -0,0 +1,1053 @@
+#include "vapor/glutil.h"
+#include "vapor/FlowRenderer.h"
+#include "vapor/SteadyVAPORVelocity.h"
+#include "vapor/UnsteadyVAPORVelocity.h"
+#include "vapor/SteadyVAPORScalar.h"
+#include "vapor/Particle.h"
+#include <iostream>
+#include <cstring>
+#include <random>
+
+#ifdef WIN32
+    #include <Processthreadsapi.h>
+#else
+    #include <sys/types.h>
+    #include <unistd.h>
+#endif
+
+#define GL_ERROR     -20
+
+using namespace VAPoR;
+
+static RendererRegistrar<FlowRenderer> registrar( FlowRenderer::GetClassType(), 
+                                                  FlowParams::GetClassType() );
+
+GLenum glCheckError_(const char *file, int line)
+{
+    GLenum errorCode;
+    while ((errorCode = glGetError()) != GL_NO_ERROR)
+    {
+        std::string error;
+        switch (errorCode)
+        {
+            case GL_INVALID_ENUM:                  error = "INVALID_ENUM"; break;
+            case GL_INVALID_VALUE:                 error = "INVALID_VALUE"; break;
+            case GL_INVALID_OPERATION:             error = "INVALID_OPERATION"; break;
+            case GL_STACK_OVERFLOW:                error = "STACK_OVERFLOW"; break;
+            case GL_STACK_UNDERFLOW:               error = "STACK_UNDERFLOW"; break;
+            case GL_OUT_OF_MEMORY:                 error = "OUT_OF_MEMORY"; break;
+            case GL_INVALID_FRAMEBUFFER_OPERATION: error = "INVALID_FRAMEBUFFER_OPERATION"; break;
+        }
+        std::cout << error << " | " << file << " (" << line << ")" << std::endl;
+    }
+    return errorCode;
+}
+#define glCheckError() glCheckError_(__FILE__, __LINE__)
+
+
+// Constructor
+FlowRenderer::FlowRenderer( const ParamsMgr*    pm,
+                      std::string&        winName,
+                      std::string&        dataSetName,
+                      std::string&        instName,
+                      DataMgr*            dataMgr )
+          : Renderer( pm,
+                      winName,
+                      dataSetName,
+                      FlowParams::GetClassType(),
+                      FlowRenderer::GetClassType(),
+                      instName,
+                      dataMgr ),
+            _velocityField     ( 9 ),
```

Comment:
Is this setting the cache size to 9 fields, and does that mean that up to 9 Grid instances might be kept in memory? Could it be reduced to 6 (3 velocity components * 2 time steps)? For big data this could be a problem.

Reply:
It is correct. Sure, I can change it to 6.